qiskit/converters/circuit_to_instruction.py

```diff
@@ -59,20 +62,26 @@ def circuit_to_instruction(circuit, parameter_map=None):
             circuit_to_instruction(circ)
     """
 
-    if parameter_map is None:
-        parameter_dict = {p: p for p in circuit.parameters}
+    if sys.version_info.major >= 3 and sys.version_info.minor >= 6:
+        parameter_dict = {}
     else:
-        parameter_dict = circuit._unroll_param_dict(parameter_map)
+        parameter_dict = OrderedDict()
 
-    if parameter_dict.keys() != circuit.parameters:
-        raise QiskitError(('parameter_map should map all circuit parameters. '
-                           'Circuit parameters: {}, parameter_map: {}').format(
-                               circuit.parameters, parameter_dict))
+    if parameter_map is None:
+        parameter_dict.update(zip(circuit.parameters, circuit.parameters))
+    else:
+        unrolled_parameter_map = circuit._unroll_param_dict(parameter_map)
+        if unrolled_parameter_map.keys() != set(circuit.parameters):
+            raise QiskitError(('parameter_map should map all circuit parameters. '
+                               'Circuit parameters: {}, parameter_map: {}').format(
+                                   circuit.parameters, parameter_dict))
+        for parameter in circuit.parameters:
+            parameter_dict[parameter] = unrolled_parameter_map[parameter]
 
     instruction = Instruction(name=circuit.name,
                               num_qubits=sum([qreg.size for qreg in circuit.qregs]),
                               num_clbits=sum([creg.size for creg in circuit.cregs]),
-                              params=sorted(parameter_dict.values(), key=lambda p: p.name))
+                              params=list(parameter_dict.values()))
```

Comment:
Is there a way we can deprecate the current behavior?

Reply:
We could add keywords and enabling sorting optionally, then by default, but I’m not sure if that’s worth the trouble. Also, in the current status of the code no sorting is presumed so I think it would be okay to change it. What do you think?

Reply:
Sort-by-name is there currently for generating the `params` list of the output gate, so a circuit like
```
qc = qk.QuantumCircuit(1)
th = Parameter('theta')
ph = Parameter('phi')
qc.u1(th, 0)
qc.u1(ph, 0)
g = qc.to_gate()
```
would generate a gate like `g(phi, theta)` currently and `g(theta, phi)` under insertion ordering.

Reply:
I changed the behaviour to be name sorted by default and otherwise the order of the circuit parameters are inherited. I.e. changed the signature to `QuantumCircuit.to_gate(parameter_map=None, sort_parameters_by_name=True)`. 

Due to yesterday's discussion I didn't yet add a deprecation warning.